In [3]:
from collections import defaultdict

# 538 October 14 Riddler
The key is this strategy. If you are ahead, you should rush; if you are behind, you should pass.

Why? Suppose you are ahead 5-0. If you rush, you will need to lose 5 more times than you win to give your lead up. However, if you pass, you will only need to lose 3 more times than you win to give your lead up. Thus, you should rush, to minimize the odds of your opponent catching you. With the reverse logic, if you are behind 0-5, you should pass, so you only need 3 more wins than losses to take the lead back.

What is less clear is whether you should pass or rush when the game is tied. Thus, we will investigate all such strategies.

### Question 1: Opponent rushes, going first.

In [35]:
scores = defaultdict(int)
scores[0] = 1
for _ in range(100):
    new_scores = defaultdict(int)
    for score in scores:
        if score <= 0:
            new_scores[score - 2] += scores[score] / 2
            new_scores[score + 2] += scores[score] / 2
        else:
            new_scores[score - 1] += scores[score] / 2
            new_scores[score + 1] += scores[score] / 2
    scores = defaultdict(int)
    for score in new_scores:
        scores[score - 1] += new_scores[score] / 2
        scores[score + 1] += new_scores[score] / 2

print('Chance of winning when passing on ties: {:.10f}'.format(sum([scores[score] for score in scores if score > 0])))

scores = defaultdict(int)
scores[0] = 1
for _ in range(100):
    new_scores = defaultdict(int)
    for score in scores:
        if score < 0:
            new_scores[score - 2] += scores[score] / 2
            new_scores[score + 2] += scores[score] / 2
        else:
            new_scores[score - 1] += scores[score] / 2
            new_scores[score + 1] += scores[score] / 2
    scores = defaultdict(int)
    for score in new_scores:
        scores[score - 1] += new_scores[score] / 2
        scores[score + 1] += new_scores[score] / 2

print('Chance of winning when rushing on ties: {:.10f}'.format(sum([scores[score] for score in scores if score > 0])))

Chance of winning when passing on ties: 0.5983729313
Chance of winning when rushing on ties: 0.5829823330


### Question 1: Opponent rushes, going second.

In [36]:
scores = defaultdict(int)
scores[0] = 1
for _ in range(100):
    new_scores = defaultdict(int)
    for score in scores:
        new_scores[score - 1] += scores[score] / 2
        new_scores[score + 1] += scores[score] / 2
    scores = defaultdict(int)
    for score in new_scores:
        if score >= 0:
            scores[score - 2] += new_scores[score] / 2
            scores[score + 2] += new_scores[score] / 2
        else:
            scores[score - 1] += new_scores[score] / 2
            scores[score + 1] += new_scores[score] / 2

print('Chance of winning when passing on ties: {:.10f}'.format(sum([scores[score] for score in scores if score < 0])))

scores = defaultdict(int)
scores[0] = 1
for _ in range(100):
    new_scores = defaultdict(int)
    for score in scores:
        new_scores[score - 1] += scores[score] / 2
        new_scores[score + 1] += scores[score] / 2
    scores = defaultdict(int)
    for score in new_scores:
        if score > 0:
            scores[score - 2] += new_scores[score] / 2
            scores[score + 2] += new_scores[score] / 2
        else:
            scores[score - 1] += new_scores[score] / 2
            scores[score + 1] += new_scores[score] / 2

print('Chance of winning when rushing on ties: {:.10f}'.format(sum([scores[score] for score in scores if score < 0])))

Chance of winning when passing on ties: 0.5958085345
Chance of winning when rushing on ties: 0.5984614273


### Question 2: Opponent passes, going first.

In [37]:
scores = defaultdict(int)
scores[0] = 1
for _ in range(100):
    new_scores = defaultdict(int)
    for score in scores:
        if score <= 0:
            new_scores[score - 2] += scores[score] / 2
            new_scores[score + 2] += scores[score] / 2
        else:
            new_scores[score - 1] += scores[score] / 2
            new_scores[score + 1] += scores[score] / 2
    scores = defaultdict(int)
    for score in new_scores:
        scores[score - 2] += new_scores[score] / 2
        scores[score + 2] += new_scores[score] / 2

print('Chance of winning when passing on ties: {:.10f}'.format(sum([scores[score] for score in scores if score > 0])))

scores = defaultdict(int)
scores[0] = 1
for _ in range(100):
    new_scores = defaultdict(int)
    for score in scores:
        if score < 0:
            new_scores[score - 2] += scores[score] / 2
            new_scores[score + 2] += scores[score] / 2
        else:
            new_scores[score - 1] += scores[score] / 2
            new_scores[score + 1] += scores[score] / 2
    scores = defaultdict(int)
    for score in new_scores:
        scores[score - 2] += new_scores[score] / 2
        scores[score + 2] += new_scores[score] / 2

print('Chance of winning when rushing on ties: {:.10f}'.format(sum([scores[score] for score in scores if score > 0])))

Chance of winning when passing on ties: 0.5370450022
Chance of winning when rushing on ties: 0.5436469571


### Question 2: Opponent passes, going second.

In [38]:
scores = defaultdict(int)
scores[0] = 1
for _ in range(100):
    new_scores = defaultdict(int)
    for score in scores:
        new_scores[score - 2] += scores[score] / 2
        new_scores[score + 2] += scores[score] / 2
    scores = defaultdict(int)
    for score in new_scores:
        if score >= 0:
            scores[score - 2] += new_scores[score] / 2
            scores[score + 2] += new_scores[score] / 2
        else:
            scores[score - 1] += new_scores[score] / 2
            scores[score + 1] += new_scores[score] / 2

print('Chance of winning when passing on ties: {:.10f}'.format(sum([scores[score] for score in scores if score < 0])))

scores = defaultdict(int)
scores[0] = 1
for _ in range(100):
    new_scores = defaultdict(int)
    for score in scores:
        new_scores[score - 2] += scores[score] / 2
        new_scores[score + 2] += scores[score] / 2
    scores = defaultdict(int)
    for score in new_scores:
        if score > 0:
            scores[score - 2] += new_scores[score] / 2
            scores[score + 2] += new_scores[score] / 2
        else:
            scores[score - 1] += new_scores[score] / 2
            scores[score + 1] += new_scores[score] / 2

print('Chance of winning when rushing on ties: {:.10f}'.format(sum([scores[score] for score in scores if score < 0])))

Chance of winning when passing on ties: 0.5528447290
Chance of winning when rushing on ties: 0.5546760108


### Question 3: Finding an optimal strategy
Four possibilities:
1. Pass on ties, always.
2. Pass on ties when going first.
3. Pass on ties when going second.
4. Rush on ties, always.

In [42]:
scores = defaultdict(int)
scores[0] = 1
for _ in range(100):
    new_scores = defaultdict(int)
    for score in scores:
        if score <= 0:
            new_scores[score - 2] += scores[score] / 2
            new_scores[score + 2] += scores[score] / 2
        else:
            new_scores[score - 1] += scores[score] / 2
            new_scores[score + 1] += scores[score] / 2
    scores = defaultdict(int)
    for score in new_scores:
        if score >= 0:
            scores[score - 2] += new_scores[score] / 2
            scores[score + 2] += new_scores[score] / 2
        else:
            scores[score - 1] += new_scores[score] / 2
            scores[score + 1] += new_scores[score] / 2

print('P1 and P2 pass on ties:')
print('P1 wins with p = {:.10f}'.format(sum([scores[score] for score in scores if score > 0])))
print('P2 wins with p = {:.10f}'.format(sum([scores[score] for score in scores if score < 0])))

scores = defaultdict(int)
scores[0] = 1
for _ in range(100):
    new_scores = defaultdict(int)
    for score in scores:
        if score <= 0:
            new_scores[score - 2] += scores[score] / 2
            new_scores[score + 2] += scores[score] / 2
        else:
            new_scores[score - 1] += scores[score] / 2
            new_scores[score + 1] += scores[score] / 2
    scores = defaultdict(int)
    for score in new_scores:
        if score > 0:
            scores[score - 2] += new_scores[score] / 2
            scores[score + 2] += new_scores[score] / 2
        else:
            scores[score - 1] += new_scores[score] / 2
            scores[score + 1] += new_scores[score] / 2

print('P1 passes, P2 rushes on ties:')
print('P1 wins with p = {:.10f}'.format(sum([scores[score] for score in scores if score > 0])))
print('P2 wins with p = {:.10f}'.format(sum([scores[score] for score in scores if score < 0])))

scores = defaultdict(int)
scores[0] = 1
for _ in range(100):
    new_scores = defaultdict(int)
    for score in scores:
        if score < 0:
            new_scores[score - 2] += scores[score] / 2
            new_scores[score + 2] += scores[score] / 2
        else:
            new_scores[score - 1] += scores[score] / 2
            new_scores[score + 1] += scores[score] / 2
    scores = defaultdict(int)
    for score in new_scores:
        if score >= 0:
            scores[score - 2] += new_scores[score] / 2
            scores[score + 2] += new_scores[score] / 2
        else:
            scores[score - 1] += new_scores[score] / 2
            scores[score + 1] += new_scores[score] / 2

print('P1 rushes, P2 passes on ties:')
print('P1 wins with p = {:.10f}'.format(sum([scores[score] for score in scores if score > 0])))
print('P2 wins with p = {:.10f}'.format(sum([scores[score] for score in scores if score < 0])))

scores = defaultdict(int)
scores[0] = 1
for _ in range(100):
    new_scores = defaultdict(int)
    for score in scores:
        if score < 0:
            new_scores[score - 2] += scores[score] / 2
            new_scores[score + 2] += scores[score] / 2
        else:
            new_scores[score - 1] += scores[score] / 2
            new_scores[score + 1] += scores[score] / 2
    scores = defaultdict(int)
    for score in new_scores:
        if score > 0:
            scores[score - 2] += new_scores[score] / 2
            scores[score + 2] += new_scores[score] / 2
        else:
            scores[score - 1] += new_scores[score] / 2
            scores[score + 1] += new_scores[score] / 2

print('P1 and P2 rush on ties:')
print('P1 wins with p = {:.10f}'.format(sum([scores[score] for score in scores if score > 0])))
print('P2 wins with p = {:.10f}'.format(sum([scores[score] for score in scores if score < 0])))

P1 and P2 pass on ties:
P1 wins with p = 0.4874372234
P2 wins with p = 0.4982643283
P1 passes, P2 rushes on ties:
P1 wins with p = 0.4844757564
P2 wins with p = 0.4989884497
P1 rushes, P2 passes on ties:
P1 wins with p = 0.4874693683
P2 wins with p = 0.4977085415
P1 and P2 rush on ties:
P1 wins with p = 0.4840207914
P2 wins with p = 0.4985234889


Since both win probs are higher for P2 if P2 rushes on ties; P2 will always rush on ties. Then P1 gains the best chance of winning by passing on ties.

Hence the final probabilities are P1 winning with probability 0.4844757564 and P2 winning with probability 0.4989884497.